### DISCUSSION TITLE
Format Function

### DISCUSSION THREAD

**Discussion by Peter-Jacob** [adesutherland/CREXX#187] [Q&A]
who?

**Comment by Peter-Jacob** [comment]
No volunteers? ok, then I will burn my fingers on it!

**Comment by adesutherland** [comment]
You have chosen unwisely! LOL

More seriously - thanks and I think we all know this is a huge task and there will be lots of debate.

I did a C version for BREXX (a nightmare), would it be useful if I extract the highlevel structure logic. It took [me] a long time to work out a structure that works. On the other hand the ANSI standards may give a clearer view? Let us know how we can help!

**Comment by Peter-Jacob** [comment]
we tried the same on the BREXX side (in the C coding), as its format, is totally different.  We gave up, decided it's not worth it.
Maybe a rexx BIF version is easier. We will see!

**Comment by Peter-Jacob** [comment]
I committed the first version of FORMAT, it just contains the easy part FORMAT(number,before,after): commit 322316355e37ac43a6999247efa8975d27747839

The missing parts supporting FORMAT(number,before,after,expp,expt) will come in part II. 
Nonetheless, I'd appreciate tests on it. Maybe we need some rounding, documentations are unclear

**Comment by adesutherland** [comment]
Thinking about it - some [many] of FORMAT's features will probably be blocked until we get the REXX numeric class done. At the moment we probably will have to stick with format for doubles, if that makes sense

**Comment by Peter-Jacob** [comment]
@adesutherland for the display with exponent we might need a special instruction (using sprint, ...) to convert a float into it, or do you see a better way?

**Comment by Peter-Jacob** [comment]
I added a FTOSX (FTSO exponent, or extended) instruction, which uses a %e print format
 ```
START_INSTRUCTION(FTOSX_REG) CALC_DISPATCH(1);
            DEBUG("TRACE - FTOSX R%llu\n", REG_IDX(1));
            prep_string_buffer(op1R, SMALLEST_STRING_BUFFER_LENGTH); // Large enough for a float
            op1R->string_length = snprintf(op1R->string_value,SMALLEST_STRING_BUFFER_LENGTH,"%e",op1R->float_value);
            op1R->string_pos = 0;
            DISPATCH;
```
I think this is good enough to start!

```
fnum=314151617.1234567
assembler ftosx fnum
say "float "fnum
```
`float 3.14152e+008`

question is of course how compatible it is over platforms and compilers

**Reply by adesutherland** [reply]
Well that is great - and printf is rather standard across platforms. And this makes me wonder if we should not introduce a printf instruction (called printf or maybe cformat)

printf output_reg format_specifier_reg num_of_args_reg (i.e. handling arguments like call)

```
sload r2, "%e"
iload r3,1
fload r4,132.24
cformat r1,r2,r3
say r1 /* Prints something like 1.3224e+002
```
We would also need to extend the ASSEMBLE instruction to handle the extra registers ... doable

And in the short term we could have a simpler instruction that only supports one argument
```
sload r2, "%e"
fload r10,132.24
cformat1 r1,r2,r10 /* note the **1** */
say r1 /* Prints something like 1.3224e+002
```

What do you think - I think you could do the cformat1 quite easily? And I can help with the more complex one.

**Comment by Peter-Jacob** [comment]
Can printf/snprintf handle dynamic formats?
e.g. printf("this is my number %d\n",mynum) would become something like:
```
myformat="%d\n"
printf("this is my number "myformat,mynum)
```
If this is possible in one way or the other, it will open new possibilities. We could have
flexible printf instruction and snprint instructions.

**Reply by adesutherland** [reply]
yes, in c printf() etc. does support dynamic formats.

I was talking about rxas instructions - but yeah there is no reason why we should not have a rexx level b printf(), if that is what you are suggesting?

**Reply by Peter-Jacob** [reply]
First of all, we need a kind snprintf rxas instruction, which returns the created string of it, printf maybe too, but we have SAY, so I am not to sure.  
it will work as you described it:
```
sload r2, "%e"
fload r4,132.24
cformat r1,r2,r4
say r1 /* Prints something like 1.3224e+002
```
iload r3,1      <-------------  why do we need this?

I don't see the necessity of an extra fourth register.

**Reply by adesutherland** [reply]
In this case we only have one argument - a single %e - (that is why I suggested the cformat1 instruction as well).

But we might have more than one argument - e.g. doing Engineering format (not directly supported by printf) might need something like
```
sload r2, "%fE%02d"
iload r3,2
fload r4,132.2
iload r5, 3
cformat r1,r2,r3
say r1 /* Says "132.2E03" */

```

**Comment by Peter-Jacob** [comment]
I am really excited, this is the instruction;
  ```
            START_INSTRUCTION(CFORMAT_REG_REG_REG) CALC_DISPATCH(3);
            DEBUG("TRACE - STRCHAR R%d,R%d,R%d\n", (int)REG_IDX(1), (int)REG_IDX(2), (int)REG_IDX(3));
            prep_string_buffer(op1R,64); // assign 64 bytes
            op1R->string_length = snprintf(op1R->string_value,64,op3R->string_value,op2R->float_value);
            op1R->string_pos = 0;
            DISPATCH;
```
the REXX:
```
rs1=""
rs2=""
float=3.1415926
f1="My float conversion (with exponent)  %e"
assembler cformat rs1,float,f1
say "Result "rs1

float=3.14
f2="simple Float conversion %f"
assembler cformat rs2,float,f2
say "Result "rs2
```
the result:
```
Result My float conversion (with exponent)  3.141593e+000
Result simple Float conversion 3.140000 (with exponent)  3.141593e+000
```
There seems to be a problem with allocating the new result buffer (parts of the old are contained). I will investigate in it. 
Overall a really nice accomplishment!

**Comment by adesutherland** [comment]
This is how set_string() works from rxvmvars.h
```
static RX_INLINE void set_string(value *v, char *value, size_t length) {
    prep_string_buffer(v,length);
    memcpy(v->string_value, value, v->string_length);
    v->string_pos = 0;
#ifndef NUTF8
    v->string_char_pos = 0;
    v->string_chars = utf8nlen(v->string_value, v->string_length); /* SLOW! */
#endif
}
```
Note the UTF stuff

Also snprintf() should be used like this (sorry for the incomplete / buggy C but hopefully you will get my point!) :-)
```
char* format_specifier; /* argument */
size_t length;
size_t buffer_size = strlen(format_specifier * 2); /* Generous guess! */
char* buffer = malloc(buffer_size);

length = snprintf(buffer, buffer_size, format_specifier, arg1, ...);
if (length > buffer_size) { /* The buffer wasn't big enough! - This is the bit you need to add */
  buffer_size = length + 1; /* +1 for null */
  free(buffer);
  buffer = malloc(buffer_size);
  length = snprintf(buffer, buffer_size, format_specifier, arg1, ...); /* Now we know the buffer is big enough */
}

/* Now set the register */
set_string(register, buffer, length);
free(buffer);
```  

But if you don't use set_string() you can avoid a string copy ...

**Comment by Peter-Jacob** [comment]
Actually I have stolen the code from FTOS and added reg2,reg3:
```
             prep_string_buffer(op1R,64); // Large enough for a float
            op1R->string_length = snprintf(op1R->string_value,64,op3R->string_value,op2R->float_value);
            op1R->string_pos = 0;
  #ifndef NUTF8
            op1R->string_char_pos = 0;
            op1R->string_chars = op1R->string_length;
  #endif
```
I thought in this case I don't need an extra set_string call, am I wrong?

**Reply by adesutherland** [reply]
In that case no ... as you have the same logic. But I do think you need to worry about a buffer overflow if the op1R->string_value is quite long ...

**Comment by Peter-Jacob** [comment]
ok, understood, I assume the snprintf sets the terminating 0x, doesn't it?

**Reply by adesutherland** [reply]
It does - but the length it returns does not include the null terminator (hence the + 1 in the malloc).

Also in ftos we have

    op1R->string_chars = op1R->string_length;

Which is fine as the float as a string is pure ASCII and has no UTF - but more generally we need 

    v->string_chars = utf8nlen(v->string_value, v->string_length);

**Comment by Peter-Jacob** [comment]
My fformat instruction works well for floats(I renamed it, as we might have iformat, sformat ...)! I had some trouble with the formatting string, as I didn't realise rexx variables are not necessarily null-terminated, which is needed for snprintf. 
Just in case, this is what you need:
           **op3R->string_value[op3R->string_length]='\0';    // terminate format string explicitly, rexx vars aren't!**
 ```
  START_INSTRUCTION(CFORMAT_REG_REG_REG) CALC_DISPATCH(3);
            DEBUG("TRACE - STRCHAR R%d,R%d,R%d\n", (int)REG_IDX(1), (int)REG_IDX(2), (int)REG_IDX(3));
            prep_string_buffer(op1R,SMALLEST_STRING_BUFFER_LENGTH); // Large enough for a float
            op3R->string_value[op3R->string_length]='\0';    // terminate format string explicitly, rexx vars aren't!
            op1R->string_length = snprintf(op1R->string_value,SMALLEST_STRING_BUFFER_LENGTH,op3R->string_value,op2R->float_value);
            op1R->string_pos = 0;
  #ifndef NUTF8
            op1R->string_char_pos = 0;
            op1R->string_chars = op1R->string_length;
  #endif
            DISPATCH;
```

**Comment by Peter-Jacob** [comment]
Format function part II commit 32f8e9a3b09d51c86d100b9b48734e0df99e97fb
it contains the formatting of the exponent. I tested with some samples of the Regina manual, which fail. I think its samples are wrong, but maybe my understanding!
So happy testing!

**Comment by rvjansen** [comment]
Hi Peter,

Thanks!

For completeness and probably superfluous: you did have a look at docs/bifs/format.rexx ?

**Comment by Peter-Jacob** [comment]
no I didn't! Have you already started with the implementation?

**Comment by rvjansen** [comment]
No, these are for reference. I used some of the ideas for the BIFs I did. This is work by Brian Marks for CRX, the verification implementation of the ANSI standard.

**Comment by Peter-Jacob** [comment]
ok, I noticed it's a "real" rexx implementation. In my proposal, I had used documentation from: IBM, Regina, and Howard Fosdicks' REXX Programmer's Reference and tried to piece it together. I am sure we have to adjust it a lot. 
Do you have a test script for Format? I'd like to use a reliable test case.
